pam/internal/adapter/nativemodel.go

```diff
@@ -325,6 +325,16 @@ func (m nativeModel) checkForPromptReplyValidity(reply string) error {
 	return nil
 }
 
+func (m nativeModel) startStringConv(style pam.Style, format string, args ...any) (pam.StringConvResponse, error) {
+	if (style == pam.TextInfo || style == pam.ErrorMsg) && len(format) > 0 &&
+		IsSSHSession(m.pamMTx) {
+		// SSH shows the prompt message just after the user name, making our
+		// UI not to look as expected, so move it to the next line.
+		format = "\n" + format
```

Comment:
This doesn't seem to work as expected, the lines are not printed after the username:

```
ssh foo@bar.onmicrosoft.com@localhost

== Local Password Authentication ==
Enter 'r' to cancel the request and go back to select the authentication method
(foo@bar.onmicrosoft.com@localhost) Enter your local password:
> 

== Authentication method selection ==
  1. Local Password Authentication
  2. Device Authentication
Or enter 'r' to go back to choose the provider
(foo@bar.onmicrosoft.com@localhost) Choose your authentication method:
> 1

== Local Password Authentication ==
Enter 'r' to cancel the request and go back to select the authentication method
(foo@bar.onmicrosoft.com@localhost) Enter your local password:
> 
```

Reply:
For what it's worth, I don't think it looks too bad that the `Enter your local password:` prompt is printed in the same line as the username.

Reply:
That was the output in noble, isn't it?

As IIRC that has a different output than jammy (or anything pre-noble, as per our sshd changes), the problem is that our CI isn't in noble (for some crazy reason, since my fork does run in noble), thus the golden files would mismatch.

So maybe it's also better to check for the version we're running in and ensure the look is consistent in both or what?

Reply:
> That was the output in noble, isn't it?

Yes, the output I pasted above is from noble.

> the problem is that our CI isn't in noble (for some crazy reason, since my fork does run in noble), thus the golden files would mismatch.

I can't really follow. Which golden files would mismatch in which case?

Reply:
So... Noble's `ssh` has a different handling of PAM messages so the output is also different (prompts aren't handled the same way).

However, maybe better to leave this change out for another branch where things can be clearer.

Reply:
> However, maybe better to leave this change out for another branch where things can be clearer.

Yeah that might be better. The current state doesn't look on Noble at least.

Reply:
Btw the golden files for the SSH test cases don't show the username prefix anywhere.

Reply:
so ideally that's the difference I'd like to see in result: https://github.com/ubuntu/authd/pull/583/commits/4bf67ed7995f23e0b2eddd4c02357d2b3929064c

But since this is anyways variable with jammy vs noble, I think it's better to wait until CI is stabilized to noble (since as per discussion with @didrocks we don't want to force to use noble machine in the workflow file)

Reply:
Commit is dropped, will handle this in an upcoming PR.

Reply:
In order not to break developers experience though, SSH tests are now running only when the host is older than noble, otherwise we'd end up with wrong golden files all the times.

It's not optimal, but there's no other way unless we switch to `runs-on: ubuntu-noble` explicitly (which imho at this point it would be better to track what we target in the PPA).

Reply:
I wonder why we're not just switching to `ubuntu-24.04` in our CI config

Reply:
I don’t think it’s a question of `jammy` vs `noble`, but rather, avoid harcoding yet in another place `ubuntu-<version>` and always follow `ubuntu-latest` unless there is a good reason to have a particular use case for it.

I don’t think in Go in particular, being on an older version is an issue for package tests (or it means we don’t isolate well enough from the platform). Ofc, end to end tests should be running in targeted releases (we have for instance e2e tests running on 22.04 and 24.04).

Reasoning for always preferring `-latest`: we have tens/hundreds of projects. If every 2 years or 6 months (depending on what we support), we need to bump them all, this is unecessary maintainance burden and we will probably forget some places to do this change.

It’s annoying that 6 more after GitHub [`ubuntu-latest` is still on 22.04](https://github.com/actions/runner-images). As you can see, the rollout will start on the [5th of December and up to mid January](https://github.com/actions/runner-images/issues/10636). If this is blocking you, maybe we should have some GitHub actions to help us there? Maybe some already exists?

Reply:
> Reasoning for always preferring -latest: we have tens/hundreds of projects. If every 2 years or 6 months (depending on what we support), we need to bump them all, this is unecessary maintainance burden and we will probably forget some places to do this change.

I can very much relate to you wanting to avoid manual maintenance steps. However, I find it not practical to avoid manual maintenance altogether, so to ensure that we don't forget to do it, we should have tooling in place which ensures that we think about it at the right time. I don't know if we currently have a process to schedule tasks which we should do after a Ubuntu release - if not, I suggest we think about ways to do that.

In this particular case, in my opinion the manual effort of searching and replacing `ubuntu-22.04` with `ubuntu-24.04` is small (when we have the tooling to schedule it), and is worth it if it means that we can still run all the tests locally on noble (which is not possible with https://github.com/ubuntu/authd/pull/583/commits/b89f5f536660c0d9649290a8e831906ee2db0dad IIUC).

Reply:
Yeah, I agree that it could be problem to maintain this for many projects, but I think for some specific cases that are under active development (like authd*) it's fine to have special cases so that we don't rely on GH to decide when we depend on a particular version.

I also guess that ideally in future we should run (at least the integration tests) in a matrix of ubuntu versions we support, maybe using properly crafted docker containers instead of on the default GH one.

Personally in order to maintain this branch I had to work in a docker container, that isn't the worst thing, but neither something we should expect anyone in the team to have to be able to check these tests.

So that said... let me know how to proceed, but I'd be in favor of doing the noble switch now, so that we don't have to re-factor this change again or force us to use older versions to fully test authd (and in an environment that we don't target so far).

Reply:
@adombeck even manual, it’s still one more steps and things that are forgotten. I have many examples of that in various teams happening.

My suggestion would be to add as part of the tooling some similar workflow than the others. We can have an action we import from the desktop-engineering repo, scanning all workflows files, and looking for:
```
   runs-on: ubuntu-22.04 // track-latest
```

And open a PR whenever ubuntu-24.04 is available (maybe the detection could be as simple as monitoring this: https://github.com/actions/runner-images/tree/main/images/ubuntu and erroring out ofc if this directory does not exists anymore), ready to be merged. We could even imagine supporting matrix too.

(the `// track-latest` is up for debate, but I think that could signal we really want to use that version). That way, even on active repo, we would be all set.

I’m happy to have a look (that doesn’t prevent changing and pretagging this), or if someone wants to beat me to it…

WDYT?

Reply:
Moved this to https://warthogs.atlassian.net/browse/UDENG-5262